proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/biometry/impl/BiometricAuthenticator.java

```diff
@@ -0,0 +1,354 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getlime.security.powerauth.biometry.impl;
+
+import android.content.Context;
+import android.content.DialogInterface;
+import android.content.pm.PackageManager;
+import android.hardware.biometrics.BiometricPrompt;
+import android.hardware.fingerprint.FingerprintManager;
+import android.os.Build;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.RequiresApi;
+import android.support.v4.app.FragmentManager;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+
+import io.getlime.security.powerauth.biometry.BiometricAuthenticationRequest;
+import io.getlime.security.powerauth.biometry.BiometricDialogResources;
+import io.getlime.security.powerauth.biometry.BiometricStatus;
+import io.getlime.security.powerauth.biometry.IBiometricKeystore;
+import io.getlime.security.powerauth.exception.PowerAuthErrorCodes;
+import io.getlime.security.powerauth.exception.PowerAuthErrorException;
+import io.getlime.security.powerauth.networking.interfaces.ICancelable;
+import io.getlime.security.powerauth.sdk.impl.CancelableTask;
+import io.getlime.security.powerauth.sdk.impl.CompositeCancelableTask;
+
+/**
+ * The {@code BiometricAuthenticator} implements {@link IBiometricAuthenticator} interface with using new
+ * {@link BiometricPrompt} facility. This implementation is automatically used on devices with
+ * Android 9.0 and newer.
+ */
+@RequiresApi(api = Build.VERSION_CODES.P)
+public class BiometricAuthenticator implements IBiometricAuthenticator {
+
+    private final @NonNull Context context;
+    private final @NonNull IBiometricKeystore keystore;
+    private final @NonNull FingerprintManager legacyFingerprintManager;
+    private byte[] alreadyProtectedKey;
+    private boolean authenticationFailedBefore;
+
+    /**
+     * Creates a new instance of {@link BiometricAuthenticator}. The authenticator object is not created when
+     * biometric hardware is not present on the device.
+     *
+     * @param context Android {@link Context} object
+     * @param keystore Object implementing {@link IBiometricKeystore} which will manage lifetime of the biometric key.
+     * @return {@link BiometricAuthenticator} or {@code null} in case that biometry is not supported on the device.
+     */
+    public static @Nullable IBiometricAuthenticator createAuthenticator(final Context context, IBiometricKeystore keystore) {
+        if (!keystore.isKeystoreReady()) {
+            return null;
+        }
+        if (!context.getPackageManager().hasSystemFeature(PackageManager.FEATURE_FINGERPRINT)) {
```

Comment:
Well, you have to put this question to some Android developers, why the `BiometricPrompt` interface is designed as much crappy as possible in "P". The Android Q fixes this by adding [BiometricManager](https://developer.android.com/reference/android/hardware/biometrics/BiometricManager), but since that, we might use this hack with cooperation with a legacy `FingerprintManager`.

More info [here](https://issuetracker.google.com/issues/109826221) or [here](https://stackoverflow.com/questions/50968732/determine-if-biometric-hardware-is-present-and-the-user-has-enrolled-biometrics/55157633)